src/strawman/collections/CollectionStrawMan6.scala

```diff
@@ -0,0 +1,980 @@
+
+    protected[this] def newBuilder = new ListBuffer[A].mapResult(_.toList)
+
+    /** Prepend element */
+    def :: [B >: A](elem: B): List[B] =  new ::(elem, this)
+
+    /** Prepend operation that avoids copying this list */
+    def ++:[B >: A](prefix: List[B]): List[B] =
+      if (prefix.isEmpty) this
+      else prefix.head :: prefix.tail ++: this
+
+    /** When concatenating with another list `xs`, avoid copying `xs` */
+    override def ++[B >: A](xs: IterableOnce[B]): List[B] = xs match {
+      case xs: List[B] => this ++: xs
+      case _ => super.++(xs)
+    }
+
+    override def className = "List"
+  }
+
+  case class :: [+A](x: A, private[collections] var next: List[A @uncheckedVariance]) // sound because `next` is used only locally
+  extends List[A] {
+    override def isEmpty = false
+    override def head = x
+    override def tail = next
+  }
+
+  case object Nil extends List[Nothing] {
+    override def isEmpty = true
+    override def head = ???
+    override def tail = ???
+  }
+
+  object List extends IterableFactory[List] {
+    def fromIterable[B](coll: Iterable[B]): List[B] = coll match {
+      case coll: List[B] => coll
+      case _ => ListBuffer.fromIterable(coll).toList
+    }
+  }
+
+  /** Concrete collection type: ListBuffer */
+  class ListBuffer[A]
+  extends Seq[A]
+     with SeqLike[A, ListBuffer]
+     with Buildable[A, ListBuffer[A]]
+     with Builder[A, ListBuffer[A]] {
+
+    private var first, last: List[A] = Nil
+    private var aliased = false
+    private var len = 0
+
+    def iterator = first.iterator
+
+    def fromIterable[B](coll: Iterable[B]) = ListBuffer.fromIterable(coll)
+
+    def apply(i: Int) = first.apply(i)
+
+    def length = len
+    override def knownSize = len
+
+    protected[this] def newBuilder = new ListBuffer[A]
+
+    private def copyElems(): Unit = {
+      val buf = ListBuffer.fromIterable(result)
+      first = buf.first
+      last = buf.last
+      aliased = false
+    }
+
+    /** Convert to list; avoids copying where possible. */
+    def toList = {
+      aliased = true
+      first
+    }
+
+    def +=(elem: A) = {
+      if (aliased) copyElems()
+      val last1 = elem :: Nil
+      last match {
+        case last: ::[A] => last.next = last1
+        case _ => first = last1
+      }
+      last = last1
+      len += 1
+      this
+    }
+
+    def result = this
+
+    override def className = "ListBuffer"
+  }
+
+  object ListBuffer extends IterableFactory[ListBuffer] {
+    def fromIterable[B](coll: Iterable[B]): ListBuffer[B] = new ListBuffer[B] ++= coll
+  }
+
+  /** Concrete collection type: ArrayBuffer */
+  class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
+  extends Seq[A]
+     with SeqLike[A, ArrayBuffer]
+     with Buildable[A, ArrayBuffer[A]]
+     with Builder[A, ArrayBuffer[A]] {
+
+    def this() = this(new Array[AnyRef](16), 0)
+
+    private var elems: Array[AnyRef] = initElems
```

Comment:
It would be great if this problem could be solved in a strawman proposal. Getting specialized arrays internally in the collection implementations would be a big improvement over the current collections library.